File: genScripts.py
#!/usr/bin/env python3
import shlex, PKSMScript, sys, glob, shutil, os

games = ["LGPE", "USUM", "SM", "ORAS", "XY", "B2W2", "BW", "HGSS", "PT", "DP", "RS", "E", "FRLG"]

def main(args):
	print("Generating scripts...")
	script_list = []
	shutil.rmtree("scripts", True)
	shutil.rmtree("build", True)
	os.mkdir("scripts")
	if os.path.exists("src/universal"):
		# shutil.copytree("src/universal", "scripts/universal")
		script_list.extend(cleanCScripts("src/universal"))
	for game in games:
		script_list.extend(generate(game))

		if os.path.exists("src/" + game.lower()):
			script_list.extend(cleanCScripts("src/" + game.lower()))
		else:
			os.mkdir("scripts/" + game.lower())

		if os.path.exists("build"):
			for f in os.listdir("build"):
				shutil.move("build/" + f, "scripts/" + game.lower())
		scriptFiles = glob.glob("*.pksm")
		for pksmFile in scriptFiles:
			shutil.move(pksmFile, "scripts/" + game.lower())
	with open("script_list.txt", "w", encoding="UTF-8") as f:
		f.write("\n".join(script_list))

def generate(game):
	path = os.path.join("src", "scripts%s.txt" % game)
	script_list = []
	if os.path.exists(path):
		with open(path, 'r', encoding="UTF-8") as pksmArgFile:
			for line in pksmArgFile:
				if not line.startswith('#'):
					line.replace('\\', '/')
					pksmArgs = PKSMScript.parser.parse_args(shlex.split(line))
					PKSMScript.main(pksmArgs)
					if pksmArgs.d:
						script_list.append(f"{game.lower()}/{pksmArgs.d}/{pksmArgs.output}.pksm")
					else:
						script_list.append(f"{game.lower()}/{pksmArgs.output}.pksm")
	return script_list

def cleanCScripts(folder):
	os.mkdir(folder.replace("src/", "scripts/"))
	scripts_list = []
	for path, _, files in os.walk(folder):
		for fullname in files:
			data = ""
			with open(os.path.join(path, fullname), 'r', encoding="UTF-8") as f:
				data = f.read()

			data = data.replace("\r\n", "\n")

			with open(os.path.join(path.replace("src/", "scripts/"), fullname), 'w', encoding="UTF-8") as f:
				f.write(data)
			scripts_list.append(f"{path[4:]}/{fullname}")
	return scripts_list

if __name__ == '__main__':
	main(sys.argv)

File: genScripts.js
#!/usr/bin/env node
/* eslint comma-dangle: ["error", "always-multiline"], no-console: "off" */
const fs = require('fs');
const { join } = require('path');
const pksmScript = require('./PKSMScript');

const games = ['usum', 'sm', 'oras', 'xy', 'b2w2', 'bw', 'hgss', 'pt', 'dp'];
let subdir; // track if scripts use -d flag

function generate(game) {
    let scriptSrc = fs.readFileSync(join('src', `scripts${game.toUpperCase()}.txt`), 'utf8');
    scriptSrc = scriptSrc.split(/\r?\n/);
    scriptSrc.forEach((l) => {
        if (l.length && l.charAt(0) !== '#') {
            subdir = l.indexOf(' -d') > -1 || subdir;
            let scriptArgs = l.replace(/\\/g, '/').match(/"[^"]+"|'[^']+'|\S+/g);
            scriptArgs = scriptArgs.map(v => v.replace(/\b'|'\b|"/g, ''));
            pksmScript(scriptArgs);
        }
    });
}

// abstracted try/catch around FS calls
function tryFSSync(op, args, errCodes) {
    try {
        return fs[`${op}Sync`](...args);
    } catch (e) {
        if (errCodes.retry && errCodes.retry.indexOf(e.code) > -1) {
            return fs[`${op}Sync`](...args);
        } else if (errCodes.skip && errCodes.skip.indexOf(e.code) === -1) {
            console.log(`There was an error trying to execute fs.${op}Sync(${args})`);
            console.error(e);
        }
    }
}

// recursive rmDir
function rmDir(dir) {
    try {
        fs.accessSync(dir);
        const dirStat = fs.statSync(dir);
        if (dirStat.isDirectory()) {
            fs.readdirSync(dir).forEach((file) => {
                const curPath = join(dir, file);
                if (fs.statSync(curPath).isDirectory()) {
                    rmDir(curPath);
                } else {
                    fs.unlinkSync(curPath);
                }
            });
            fs.rmdirSync(dir);
        } else {
            fs.unlinkSync(dir);
        }
    } catch (e) {
        if (e.code !== 'ENOENT') {
            console.log(`There was an error accessing or deleting ${dir}`);
            console.error(e);
        }
    }
}

function genScripts(/* args */) {
    console.log('Generating scripts...');
    const srcContents = fs.readdirSync('src');

    // empty existing /build and /scripts directories
    rmDir('scripts');
    rmDir('build');
    tryFSSync('mkdir', ['scripts'], { retry: ['EPERM'] });

    const cScripts = tryFSSync('readdir', [join('src', 'universal')], { skip: ['ENOENT'] });
    if (cScripts) {
        fs.mkdirSync(join('scripts', 'universal'));
        cScripts.forEach((f) => {
            if (f.slice(-2) === '.c') {
                fs.copyFileSync(join('src', 'universal', f), join('scripts', 'universal', f));
            }
        });
    }

    games.forEach((game) => {
        // remake game directory
        // console.log(`Compiling scripts for ${game}`);
        const gameDir = join('scripts', game);
        tryFSSync('mkdir', [gameDir], { retry: ['ENOENT', 'EPERM'] });

        subdir = false;
        generate(game); // compile scripts for game

        // move scripts
        if (subdir) { // scripts with -d flag
            fs.readdirSync('build').forEach((v) => {
                const files = fs.readdirSync(join('build', v));
                if (files.length > 0) {
                    fs.mkdirSync(join(gameDir, v));
                }
                files.forEach((f) => {
                    tryFSSync('rename', [join('build', v, f), join(gameDir, v, f)], { retry: ['EPERM'] });
                });
            });
        }
        if (srcContents.indexOf(game) > -1) {
            fs.readdirSync(join('src', game)).forEach((v) => {
                if (v.slice(-2) === '.c') { // game-specific picoC scripts
                    fs.copyFileSync(join('src', game, v), join(gameDir, v));
                }
            });
        }
        fs.readdirSync('.').forEach((v) => {
            if (v.slice(-5) === '.pksm') {
                fs.renameSync(v, join(gameDir, v));
            }
        });
    });

    console.log('Finished compiling scripts');
}

module.exports = genScripts;

// execute if called directly from command line
if (require.main === module) {
    genScripts(/* process.argv */);
}
